efficientdet/iou_utils.py

```diff
@@ -0,0 +1,186 @@
+    enclose_height = tf.maximum(zero, enclose_ymax - enclose_ymin)
+    enclose_area = enclose_width * enclose_height
+    giou_v = iou_v - tf.math.divide_no_nan(
+        (enclose_area - union_area), enclose_area)
+    return giou_v
+
+  assert iou_type in ('diou', 'ciou')
+  p_center = tf.stack([(p_ymin + p_ymax) / 2, (p_xmin + p_xmax) / 2])
+  t_center = tf.stack([(t_ymin + t_ymax) / 2, (t_xmin + t_xmax) / 2])
+  euclidean = tf.linalg.norm(t_center - p_center)
+  diag_length = tf.linalg.norm(
+      [enclose_ymax - enclose_ymin, enclose_xmax - enclose_xmin])
+  diou_v = iou_v - tf.math.divide_no_nan(euclidean**2, diag_length**2)
+  if iou_type == 'diou':  # diou is the distance iou.
+    return diou_v
+
+  assert iou_type == 'ciou'
+  v = _get_v(p_height, p_width, t_height, t_width)
+  alpha = tf.math.divide_no_nan(v, ((1 - iou_v) + v))
+  return diou_v - alpha * v  # the last one is ciou.
+
+
+def iou_loss(pred_boxes: FloatType,
+             target_boxes: FloatType,
+             iou_type: Text = 'iou') -> tf.Tensor:
+  """A unified interface for computing various IoU losses.
+
+  Let B and B_gt denotes the pred_box and B_gt is the target box (ground truth):
+
+    IoU = |B & B_gt| / |B | B_gt|
+
+    GIoU = IoU - |C - B U B_gt| / C, where C is the smallest box covering B and
+    B_gt.
+
+    DIoU = IoU - E(B, B_gt)^2 / c^2, E is the Euclidean distance of the center
+    points of B and B_gt, and c is the diagonal length of the smallest box
+    covering the two boxes
+
+    CIoU = IoU - DIoU - a * v, where a is a positive trade-off parameter, and
+    v measures the consistency of aspect ratio:
+      v = (arctan(w_gt / h_gt) - arctan(w / h)) * 4 / pi^2
+    where (w_gt, h_gt) and (w, h) are the width and height of the target and
+    predicted box respectively.
+
+  The returned loss is computed as 1 - one of {IoU, GIoU, DIoU, CIoU}.
+
+  Args:
+    pred_boxes: predicted boxes, with coordinate [y_min, x_min, y_max, x_max]*.
+      It can be multiple anchors, with each anchor box has four coordinates.
+    target_boxes: target boxes, with coordinate [y_min, x_min, y_max, x_max]*.
+      It can be multiple anchors, with each anchor box has four coordinates.
+    iou_type: one of ['iou', 'ciou', 'diou', 'giou'].
+
+  Returns:
+    IoU loss float `Tensor`.
+  """
+  if iou_type not in ('iou', 'ciou', 'diou', 'giou'):
+    raise ValueError(
+        'Unknown loss_type {}, not iou/ciou/diou/giou'.format(iou_type))
+
+  pred_boxes = tf.convert_to_tensor(pred_boxes, tf.float32)
+  target_boxes = tf.cast(target_boxes, pred_boxes.dtype)
+
+  # t_ denotes target boxes and p_ denotes predicted boxes: (y, x, y_max, x_max)
+  pred_boxes_list = tf.unstack(pred_boxes, None, axis=-1)
+  target_boxes_list = tf.unstack(target_boxes, None, axis=-1)
+  assert len(pred_boxes_list) == len(target_boxes_list)
+  assert len(pred_boxes_list) % 4 == 0
+
+  iou_loss_list = []
+  for i in range(0, len(pred_boxes_list), 4):
```

Comment:
Reshape to [N,9,4], then use matrix calculate?

Reply:
Good idea. I will first check this in and factorize it with matrix ops later (since this is not part of the inference graph, the slowness might be acceptable).

BTW, since I submitted the change with your email/name, it also needs your consent. Could you help reply "@googlebot I consent"? Thanks!